src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/Address.cs

```diff
@@ -0,0 +1,116 @@
+using System.Threading.Tasks;
+using Dfc.CourseDirectory.Core.Search;
+using Dfc.CourseDirectory.Core.Search.Models;
+using Dfc.CourseDirectory.Core.Validation;
+using Dfc.CourseDirectory.Core.Validation.VenueValidation;
+using FluentValidation;
+using FluentValidation.Results;
+using FormFlow;
+using MediatR;
+using OneOf;
+using OneOf.Types;
+
+namespace Dfc.CourseDirectory.WebV2.Features.Venues.AddVenue.Address
+{
+    public class Query : IRequest<Command>
+    {
+    }
+
+    public class Command : IRequest<OneOf<ModelWithErrors<Command>, Success>>
+    {
+        public string AddressLine1 { get; set; }
+        public string AddressLine2 { get; set; }
+        public string Town { get; set; }
+        public string County { get; set; }
+        public string Postcode { get; set; }
+    }
+
+    public class Handler :
+        IRequestHandler<Query, Command>,
+        IRequestHandler<Command, OneOf<ModelWithErrors<Command>, Success>>
+    {
+        private readonly JourneyInstance<AddVenueJourneyModel> _journeyInstance;
+        private readonly ISearchClient<Onspd> _onspdSearchClient;
+
+        public Handler(JourneyInstanceProvider journeyInstanceProvider, ISearchClient<Onspd> onspdSearchClient)
+        {
+            _journeyInstance = journeyInstanceProvider.GetInstance<AddVenueJourneyModel>();
+            _onspdSearchClient = onspdSearchClient;
+        }
+
+        public Task<Command> Handle(Query request, CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            return Task.FromResult(new Command()
+            {
+                AddressLine1 = _journeyInstance.State.AddressLine1,
+                AddressLine2 = _journeyInstance.State.AddressLine2,
+                Town = _journeyInstance.State.Town,
+                County = _journeyInstance.State.County,
+                Postcode = _journeyInstance.State.Postcode
+            });
+        }
+
+        public async Task<OneOf<ModelWithErrors<Command>, Success>> Handle(Command request, CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            var validator = new CommandValidator();
+            var validationResult = await validator.ValidateAsync(request);
+
+            if (!validationResult.IsValid)
+            {
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
+            var onspdSearchResult = await _onspdSearchClient.Search(new OnspdSearchQuery() { Postcode = request.Postcode });
+
+            if (onspdSearchResult.Items.Count == 0)
+            {
+                validationResult = new ValidationResult(new[]
+                {
+                    new ValidationFailure(nameof(request.Postcode), "Enter a valid postcode")
+                });
+
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
+            var onspdPostcodeRecord = onspdSearchResult.Items.Single();
```

Comment:
In fact, you could move this above the previous condition and check for null to determine results returns